contributors/design-proposals/sig-storage/volume-topology-scheduling.md

```diff
@@ -0,0 +1,580 @@
+by existing PVs or dynamically provisioned PVs that are
+topologically-constrained to the Node.
+```
+MatchUnboundPVCs(pod *v1.Pod, node *v1.Node) (canBeBound bool, err error)
+```
+1. If all the Pod’s PVCs are bound, return true.
+2. Otherwise try to find matching PVs for all of the unbound PVCs in order of
+decreasing requested capacity.
+3. Walk through all the PVs.
+4. Find best matching PV for the PVC where PV topology is satisfied by the Node.
+5. Temporarily cache this PV in the PVC object, keyed by Node, for fast
+processing later in the priority and bind functions.
+6. Return true if all PVCs are matched.
+7. If there are still unmatched PVCs, check if dynamic provisioning is possible.
+For this alpha phase, the provisioner is not topology aware, so the predicate
+will just return true if there is a provisioner specified (internal or external).
+8. Otherwise return false.
+
+TODO: caching format and details
+
+##### Priority
+After all the predicates run, there is a reduced set of Nodes that can fit a
+Pod. A new priority function will rank the remaining nodes based on the
+unbound PVCs and their matching PVs.
+```
+PrioritizeUnboundPVCs(pod *v1.Pod, filteredNodes HostPriorityList) (rankedNodes HostPriorityList, err error) 
+```
+1. For each Node, get the cached PV matches for the Pod’s PVCs.
+2. Compute a priority score for the Node using the following factors:
+    1. How close the PVC’s requested capacity and PV’s capacity are.
+    2. Matching static PVs is preferred over dynamic provisioning.
+
+TODO (beta): figure out weights and exact calculation
+
+##### Assume
+Once all the predicates and priorities have run, then the scheduler picks a
+Node.  Then we can bind or provision PVCs for that Node.  For better scheduler
+performance, we’ll assume that the binding will likely succeed, and update the
+PV cache first.  Then the actual binding API update will be made
+asynchronously, and the scheduler can continue processing other Pods.
+
+For the alpha phase, the AssumePVCs function will be directly called by the
+scheduler.  We’ll consider creating a generic scheduler interface in a
+subsequent phase.
+
+```
+AssumePVCs(pod *v1.Pod, node *v1.Node) (pvcBindingRequired bool, err error)
+```
+1. If all the Pod’s PVCs are bound, return false.
+2. For static PV binding:
+    1. Get the cached matching PVs for the PVCs on that Node.
+    2. Validate the actual PV state.
+    3. Mark PV.ClaimRef in the PV cache.
+3. For in-tree and external dynamic provisioning:
+    1. Nothing.
+4. Return true.
+
+
+##### Bind
+If AssumePVCs returns pvcBindingRequired, then the BindPVCs function is called
+as a go routine.  Otherwise, we can continue with assuming and binding the Pod
+to the Node.
+
+For the alpha phase, the BindUnboundPVCs function will be directly called by the
+scheduler.  We’ll consider creating a generic scheduler interface in a subsequent
+phase.
+
+```
+BindUnboundPVCs(pod *v1.Pod, node *v1.Node) (err error)
+```
```

Comment:
Do we check again if PVC constraints are still met? Can we clarify how this function accesses PV cache - because from what I see it only get pod and hence it must only have pvc name.